--- a/net/minecraft/entity/projectile/FishingBobberEntity.java
+++ b/net/minecraft/entity/projectile/FishingBobberEntity.java
@@ -5,7 +5,6 @@
 import javax.annotation.Nullable;
 import net.minecraft.advancement.criterion.Criterions;
 import net.minecraft.block.Block;
-import net.minecraft.block.Blocks;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.MovementType;
@@ -15,7 +14,6 @@
 import net.minecraft.server.network.ServerPlayerEntity;
 import net.minecraft.fluid.FluidState;
 import net.minecraft.item.ItemStack;
-import net.minecraft.item.Items;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.Packet;
 import net.minecraft.entity.data.TrackedData;
@@ -42,8 +40,9 @@
 import net.minecraft.world.loot.LootTables;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
-import net.minecraft.entity.ProjectileUtil;
 
+import net.minecraft;
+import net.minecraft.entity.ProjectileUtil;
 
 public class FishingBobberEntity extends Entity {
    private static final TrackedData<Integer> HOOK_ENTITY_ID = DataTracker.registerData(FishingBobberEntity.class, TrackedDataHandlerRegistry.INTEGER);
@@ -212,8 +211,8 @@
    private boolean method_6959() {
       ItemStack itemstack = this.owner.getMainHandStack();
       ItemStack itemstack1 = this.owner.getOffHandStack();
-      boolean flag = itemstack.getItem() == Items.FISHING_ROD;
-      boolean flag1 = itemstack1.getItem() == Items.FISHING_ROD;
+      boolean flag = itemstack.getItem() instanceof net.minecraft.item.FishingRodItem;
+      boolean flag1 = itemstack1.getItem() instanceof net.minecraft.item.FishingRodItem;
       if (!this.owner.removed && this.owner.isAlive() && (flag || flag1) && !(this.squaredDistanceTo(this.owner) > 1024.0D)) {
          return false;
       } else {
@@ -297,7 +296,7 @@
             double d1 = (double)((float)MathHelper.floor(this.getBoundingBox().minY) + 1.0F);
             double d2 = this.z + (double)(f2 * (float)this.field_7172 * 0.1F);
             Block block = serverworld.getBlockState(new BlockPos(d0, d1 - 1.0D, d2)).getBlock();
-            if (block == Blocks.WATER) {
+            if (serverworld.getBlockState(new BlockPos((int)d0, (int)d1 - 1, (int)d2)).getMaterial() == net.minecraft.block.Material.WATER) {
                if (this.random.nextFloat() < 0.15F) {
                   serverworld.spawnParticles(ParticleTypes.BUBBLE, d0, d1 - (double)0.1F, d2, 1, (double)f1, 0.1D, (double)f2, 0.0D);
                }
@@ -334,7 +333,7 @@
             double d5 = (double)((float)MathHelper.floor(this.getBoundingBox().minY) + 1.0F);
             double d6 = this.z + (double)(MathHelper.cos(f6) * f7 * 0.1F);
             Block block1 = serverworld.getBlockState(new BlockPos(d4, d5 - 1.0D, d6)).getBlock();
-            if (block1 == Blocks.WATER) {
+            if (serverworld.getBlockState(new BlockPos(d4, d5 - 1.0D, d6)).getMaterial() == net.minecraft.block.Material.WATER) {
                serverworld.spawnParticles(ParticleTypes.SPLASH, d4, d5, d6, 2 + this.random.nextInt(2), (double)0.1F, 0.0D, (double)0.1F, 0.0D);
             }
          }
@@ -359,6 +358,7 @@
    public int method_6957(ItemStack p_146034_1_) {
       if (!this.world.isClient && this.owner != null) {
          int i = 0;
+         net.minecraftforge.event.entity.player.ItemFishedEvent event = null;
          if (this.hookedEntity != null) {
             this.method_6954();
             Criterions.FISHING_ROD_HOOKED.handle((ServerPlayerEntity)this.owner, p_146034_1_, this, Collections.emptyList());
@@ -366,8 +366,15 @@
             i = this.hookedEntity instanceof ItemEntity ? 3 : 5;
          } else if (this.field_7173 > 0) {
             LootContext.Builder lootcontext$builder = (new LootContext.Builder((ServerWorld)this.world)).put(LootContextParameters.POSITION, new BlockPos(this)).put(LootContextParameters.TOOL, p_146034_1_).setRandom(this.random).setLuck((float)this.lureLevel + this.owner.getLuck());
+            lootcontext$builder.put(LootContextParameters.KILLER_ENTITY, this.owner).put(LootContextParameters.THIS_ENTITY, this);
             LootSupplier loottable = this.world.getServer().getLootManager().getSupplier(LootTables.FISHING_GAMEPLAY);
             List<ItemStack> list = loottable.getDrops(lootcontext$builder.build(LootContextTypes.FISHING));
+            event = new net.minecraftforge.event.entity.player.ItemFishedEvent(list, this.field_7176 ? 2 : 1, this);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+            if (event.isCanceled()) {
+               this.remove();
+               return event.getRodDamage();
+            }
             Criterions.FISHING_ROD_HOOKED.handle((ServerPlayerEntity)this.owner, p_146034_1_, this, list);
 
             for(ItemStack itemstack : list) {
@@ -392,7 +399,7 @@
          }
 
          this.remove();
-         return i;
+         return event == null ? i : event.getRodDamage();
       } else {
          return 0;
       }
@@ -418,8 +425,9 @@
       return false;
    }
 
-   public void remove() {
-      super.remove();
+   @Override
+   public void remove(boolean keepData) {
+      super.remove(keepData);
       if (this.owner != null) {
          this.owner.fishHook = null;
       }
