--- a/net/minecraft/entity/projectile/ShulkerBulletEntity.java
+++ b/net/minecraft/entity/projectile/ShulkerBulletEntity.java
@@ -31,8 +31,8 @@
 import net.minecraft.server.world.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
-import net.minecraft.entity.ProjectileUtil;
 
+import net.minecraft.entity.ProjectileUtil;
 
 public class ShulkerBulletEntity extends Entity {
    private LivingEntity owner;
@@ -253,7 +253,7 @@
             }
 
             HitResult raytraceresult = ProjectileUtil.getCollision(this, true, false, this.owner, RayTraceContext.ShapeType.COLLIDER);
-            if (raytraceresult.getType() != HitResult.Type.MISS) {
+            if (raytraceresult.getType() != HitResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
                this.onHit(raytraceresult);
             }
          }
