--- a/net\minecraft\entity\projectile\ShulkerBulletEntity.java
+++ b/net\minecraft\entity\projectile\ShulkerBulletEntity.java
@@ -256,7 +256,7 @@
          }
 
          HitResult raytraceresult = ProjectileUtil.getCollision(this, true, false, this.owner, RayTraceContext.ShapeType.field_17558);
-         if (raytraceresult.getType() != HitResult.Type.field_1333) {
+            if (raytraceresult.getType() != HitResult.Type.field_1333 && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
             this.onHit(raytraceresult);
          }
       }
