--- a/net/minecraft/entity/projectile/SmallFireballEntity.java
+++ b/net/minecraft/entity/projectile/SmallFireballEntity.java
@@ -10,7 +10,6 @@
 import net.minecraft.util.hit.BlockHitResult;
 import net.minecraft.util.hit.EntityHitResult;
 import net.minecraft.util.hit.HitResult;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 
 public class SmallFireballEntity extends AbstractFireballEntity {
@@ -40,7 +39,7 @@
                   entity.method_20803(i);
                }
             }
-         } else if (this.owner == null || !(this.owner instanceof MobEntity) || this.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+         } else if (this.owner == null || !(this.owner instanceof MobEntity) || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.owner)) {
             BlockHitResult blockraytraceresult = (BlockHitResult)result;
             BlockPos blockpos = blockraytraceresult.getBlockPos().offset(blockraytraceresult.getSide());
             if (this.world.isAir(blockpos)) {
