--- a/net/minecraft/entity/player/PlayerInventory.java
+++ b/net/minecraft/entity/player/PlayerInventory.java
@@ -216,7 +216,8 @@
       int i = p_191973_2_.getCount();
       ItemStack itemstack = this.getInvStack(p_191973_1_);
       if (itemstack.isEmpty()) {
-         itemstack = new ItemStack(item, 0);
+         itemstack = p_191973_2_.copy(); // Forge: Replace Item clone above to preserve item capabilities when picking the item up.
+         itemstack.setCount(0);
          if (p_191973_2_.hasTag()) {
             itemstack.setTag(p_191973_2_.getTag().copy());
          }
@@ -267,7 +268,7 @@
             }
          }
       }
-
+      armor.forEach(e -> e.onArmorTick(player.world, player));
    }
 
    public boolean insertStack(ItemStack itemStackIn) {
@@ -320,6 +321,8 @@
          } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.create(throwable, "Adding item to inventory");
             CrashReportSection crashreportcategory = crashreport.addElement("Item being added");
+            crashreportcategory.add("Registry Name", () -> String.valueOf(p_191971_2_.getItem().getRegistryName()));
+            crashreportcategory.add("Item Class", () -> p_191971_2_.getItem().getClass().getName());
             crashreportcategory.add("Item ID", Item.getRawId(p_191971_2_.getItem()));
             crashreportcategory.add("Item data", p_191971_2_.getDamage());
             crashreportcategory.add("Item name", () -> {
