--- a/net/minecraft/potion/Potion.java
+++ b/net/minecraft/potion/Potion.java
@@ -5,10 +5,10 @@
 import javax.annotation.Nullable;
 import net.minecraft.util.Identifier;
 import net.minecraft.util.registry.Registry;
-import net.minecraft.entity.effect.StatusEffectInstance;
 
+import net.minecraft.entity.effect.StatusEffectInstance;
 
-public class Potion {
+public class Potion extends net.minecraftforge.registries.ForgeRegistryEntry<Potion> {
    private final String name;
    private final ImmutableList<StatusEffectInstance> effectList;
 
