--- a/net/minecraft/client/resource/ClientResourcePackContainer.java
+++ b/net/minecraft/client/resource/ClientResourcePackContainer.java
@@ -23,8 +23,13 @@
    @Nullable
    private Identifier iconId;
 
+   @Deprecated
    public ClientResourcePackContainer(String p_i48113_1_, boolean p_i48113_2_, Supplier<ResourcePack> p_i48113_3_, ResourcePack p_i48113_4_, PackResourceMetadata p_i48113_5_, ResourcePackContainer.InsertionPosition p_i48113_6_) {
-      super(p_i48113_1_, p_i48113_2_, p_i48113_3_, p_i48113_4_, p_i48113_5_, p_i48113_6_);
+	  this(p_i48113_1_, p_i48113_2_, p_i48113_3_, p_i48113_4_, p_i48113_5_, p_i48113_6_, false);
+   }
+
+   public ClientResourcePackContainer(String p_i48113_1_, boolean p_i48113_2_, Supplier<ResourcePack> p_i48113_3_, ResourcePack p_i48113_4_, PackResourceMetadata p_i48113_5_, ResourcePackContainer.InsertionPosition p_i48113_6_, boolean hidden) {
+      super(p_i48113_1_, p_i48113_2_, p_i48113_3_, p_i48113_4_, p_i48113_5_, p_i48113_6_, hidden);
       NativeImage nativeimage = null;
 
       try (InputStream inputstream = p_i48113_4_.openRoot("pack.png")) {
@@ -36,8 +41,13 @@
       this.icon = nativeimage;
    }
 
+   @Deprecated
    public ClientResourcePackContainer(String p_i48114_1_, boolean p_i48114_2_, Supplier<ResourcePack> p_i48114_3_, Text p_i48114_4_, Text p_i48114_5_, ResourcePackCompatibility p_i48114_6_, ResourcePackContainer.InsertionPosition p_i48114_7_, boolean p_i48114_8_, @Nullable NativeImage p_i48114_9_) {
-      super(p_i48114_1_, p_i48114_2_, p_i48114_3_, p_i48114_4_, p_i48114_5_, p_i48114_6_, p_i48114_7_, p_i48114_8_);
+      this(p_i48114_1_, p_i48114_2_, p_i48114_3_, p_i48114_4_, p_i48114_5_, p_i48114_6_, p_i48114_7_, p_i48114_8_, p_i48114_9_, false);
+   }
+
+   public ClientResourcePackContainer(String p_i48114_1_, boolean p_i48114_2_, Supplier<ResourcePack> p_i48114_3_, Text p_i48114_4_, Text p_i48114_5_, ResourcePackCompatibility p_i48114_6_, ResourcePackContainer.InsertionPosition p_i48114_7_, boolean p_i48114_8_, @Nullable NativeImage p_i48114_9_, boolean hidden) {
+      super(p_i48114_1_, p_i48114_2_, p_i48114_3_, p_i48114_4_, p_i48114_5_, p_i48114_6_, p_i48114_7_, p_i48114_8_, hidden);
       this.icon = p_i48114_9_;
    }
 
