--- a/net/minecraft/entity/boss/WitherEntity.java
+++ b/net/minecraft/entity/boss/WitherEntity.java
@@ -45,7 +45,6 @@
 import net.minecraft.text.Text;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.explosion.Explosion;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
@@ -205,7 +204,7 @@
       if (this.getInvulTimer() > 0) {
          int j1 = this.getInvulTimer() - 1;
          if (j1 <= 0) {
-            Explosion.DestructionType explosion$mode = this.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING) ? Explosion.DestructionType.DESTROY : Explosion.DestructionType.NONE;
+            Explosion.DestructionType explosion$mode = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this) ? Explosion.DestructionType.DESTROY : Explosion.DestructionType.NONE;
             this.world.createExplosion(this, this.x, this.y + (double)this.getStandingEyeHeight(), this.z, 7.0F, false, explosion$mode);
             this.world.playGlobalEvent(1023, new BlockPos(this), 0);
          }
@@ -280,7 +279,7 @@
 
          if (this.field_7082 > 0) {
             --this.field_7082;
-            if (this.field_7082 == 0 && this.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+            if (this.field_7082 == 0 && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this)) {
                int i1 = MathHelper.floor(this.y);
                int l1 = MathHelper.floor(this.x);
                int i2 = MathHelper.floor(this.z);
@@ -294,7 +293,7 @@
                         int l = i2 + l2;
                         BlockPos blockpos = new BlockPos(i3, k, l);
                         BlockState blockstate = this.world.getBlockState(blockpos);
-                        if (canDestroy(blockstate)) {
+                        if (blockstate.canEntityDestroy(this.world, blockpos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, blockstate)) {
                            flag = this.world.breakBlock(blockpos, true) || flag;
                         }
                      }
@@ -315,6 +314,7 @@
       }
    }
 
+   @Deprecated //Forge: DO NOT USE use BlockState.canEntityDestroy
    public static boolean canDestroy(BlockState blockIn) {
       return !blockIn.isAir() && !BlockTags.WITHER_IMMUNE.contains(blockIn.getBlock());
    }
