--- a/net\minecraft\item\AxeItem.java
+++ b/net\minecraft\item\AxeItem.java
@@ -21,7 +21,7 @@
    protected static final Map<Block, Block> STRIPPED_BLOCKS = (new Builder<Block, Block>()).put(Blocks.field_10126, Blocks.field_10250).put(Blocks.field_10431, Blocks.field_10519).put(Blocks.field_10178, Blocks.field_10374).put(Blocks.field_10010, Blocks.field_10244).put(Blocks.field_9999, Blocks.field_10103).put(Blocks.field_10533, Blocks.field_10622).put(Blocks.field_10307, Blocks.field_10204).put(Blocks.field_10511, Blocks.field_10366).put(Blocks.field_10303, Blocks.field_10084).put(Blocks.field_10306, Blocks.field_10254).put(Blocks.field_10155, Blocks.field_10558).put(Blocks.field_10037, Blocks.field_10436).build();
 
    public AxeItem(ToolMaterial tier, float attackDamageIn, float attackSpeedIn, Item.Settings builder) {
-      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder);
+      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder.addToolType(net.minecraftforge.common.ToolType.AXE, tier.getMiningLevel()));
    }
 
    public float getMiningSpeed(ItemStack stack, BlockState state) {
