--- a/net/minecraft/item/ItemStack.java
+++ b/net/minecraft/item/ItemStack.java
@@ -63,11 +63,11 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+
 import net.minecraft.util.Rarity;
 import net.minecraft.util.UseAction;
 
-
-public final class ItemStack {
+public final class ItemStack extends net.minecraftforge.common.capabilities.CapabilityProvider<ItemStack> implements net.minecraftforge.common.extensions.IForgeItemStack {
    private static final Logger LOGGER = LogManager.getLogger();
    public static final ItemStack EMPTY = new ItemStack((Item)null);
    public static final DecimalFormat MODIFIER_FORMAT = createModifierFormat();
@@ -83,6 +83,9 @@
    private CachedBlockPosition lastPlaceOnPos;
    private boolean lastPlaceOnResult;
 
+   private net.minecraftforge.registries.IRegistryDelegate<Item> delegate;
+   private CompoundTag capNBT;
+
    private static DecimalFormat createModifierFormat() {
       DecimalFormat decimalformat = new DecimalFormat("#.##");
       decimalformat.setDecimalFormatSymbols(DecimalFormatSymbols.getInstance(Locale.ROOT));
@@ -93,10 +96,14 @@
       this(itemIn, 1);
    }
 
-   public ItemStack(ItemConvertible itemIn, int count) {
+   public ItemStack(ItemConvertible itemIn, int count){ this(itemIn, count, null); }
+   public ItemStack(ItemConvertible itemIn, int count, @Nullable CompoundTag capNBT) {
+      super(ItemStack.class);
+      this.capNBT = capNBT;
       this.item = itemIn == null ? null : itemIn.asItem();
       this.count = count;
       this.updateEmptyState();
+      this.forgeInit();
    }
 
    private void updateEmptyState() {
@@ -105,6 +112,8 @@
    }
 
    private ItemStack(CompoundTag compound) {
+      super(ItemStack.class);
+      this.capNBT = compound.containsKey("ForgeCaps") ? compound.getCompound("ForgeCaps") : null;
       this.item = Registry.ITEM.get(new Identifier(compound.getString("id")));
       this.count = compound.getByte("Count");
       if (compound.containsKey("tag", 10)) {
@@ -117,6 +126,7 @@
       }
 
       this.updateEmptyState();
+      this.forgeInit();
    }
 
    public static ItemStack fromTag(CompoundTag compound) {
@@ -131,7 +141,7 @@
    public boolean isEmpty() {
       if (this == EMPTY) {
          return true;
-      } else if (this.getItem() != null && this.getItem() != Items.AIR) {
+      } else if (this.getItemRaw() != null && this.getItemRaw() != Items.AIR) {
          return this.count <= 0;
       } else {
          return true;
@@ -147,10 +157,19 @@
    }
 
    public Item getItem() {
-      return this.empty ? Items.AIR : this.item;
+      return this.empty || this.delegate == null ? Items.AIR : this.delegate.get();
    }
 
    public ActionResult useOnBlock(ItemUsageContext context) {
+      if (!context.world.isClient) return net.minecraftforge.common.ForgeHooks.onPlaceItemIntoWorld(context);
+      return onItemUse(context, (c) -> getItem().useOnBlock(context));
+   }
+
+   public ActionResult onItemUseFirst(ItemUsageContext context) {
+      return onItemUse(context, (c) -> getItem().onItemUseFirst(this, context));
+   }
+
+   private ActionResult onItemUse(ItemUsageContext context, java.util.function.Function<ItemUsageContext, ActionResult> callback) {
       PlayerEntity playerentity = context.getPlayer();
       BlockPos blockpos = context.getBlockPos();
       CachedBlockPosition cachedblockinfo = new CachedBlockPosition(context.getWorld(), blockpos, false);
@@ -158,7 +177,7 @@
          return ActionResult.PASS;
       } else {
          Item item = this.getItem();
-         ActionResult actionresulttype = item.useOnBlock(context);
+         ActionResult actionresulttype = callback.apply(context);
          if (playerentity != null && actionresulttype == ActionResult.SUCCESS) {
             playerentity.incrementStat(Stats.USED.getOrCreateStat(item));
          }
@@ -186,12 +205,15 @@
       if (this.tag != null) {
          nbt.put("tag", this.tag);
       }
-
+      CompoundTag cnbt = this.serializeCaps();
+      if (cnbt != null && !cnbt.isEmpty()) {
+         nbt.put("ForgeCaps", cnbt);
+      }
       return nbt;
    }
 
    public int getMaxCount() {
-      return this.getItem().getMaxCount();
+      return this.getItem().getItemStackLimit(this);
    }
 
    public boolean isStackable() {
@@ -199,7 +221,7 @@
    }
 
    public boolean isDamageable() {
-      if (!this.empty && this.getItem().getMaxDamage() > 0) {
+      if (!this.empty && this.getItem().getMaxDamage(this) > 0) {
          CompoundTag compoundnbt = this.getTag();
          return compoundnbt == null || !compoundnbt.getBoolean("Unbreakable");
       } else {
@@ -208,19 +230,19 @@
    }
 
    public boolean isDamaged() {
-      return this.isDamageable() && this.getDamage() > 0;
+      return this.isDamageable() && getItem().isDamaged(this);
    }
 
    public int getDamage() {
-      return this.tag == null ? 0 : this.tag.getInt("Damage");
+      return this.getItem().getDamage(this);
    }
 
    public void setDamage(int damage) {
-      this.getOrCreateTag().putInt("Damage", Math.max(0, damage));
+      this.getItem().setDamage(this, damage);
    }
 
    public int getMaxDamage() {
-      return this.getItem().getMaxDamage();
+      return this.getItem().getMaxDamage(this);
    }
 
    public boolean damage(int amount, Random rand, @Nullable ServerPlayerEntity damager) {
@@ -256,6 +278,7 @@
    public <T extends LivingEntity> void damage(int amount, T entityIn, Consumer<T> onBroken) {
       if (!entityIn.world.isClient && (!(entityIn instanceof PlayerEntity) || !((PlayerEntity)entityIn).abilities.creativeMode)) {
          if (this.isDamageable()) {
+            amount = this.getItem().damageItem(this, amount, entityIn, onBroken);
             if (this.damage(amount, entityIn.getRand(), entityIn instanceof ServerPlayerEntity ? (ServerPlayerEntity)entityIn : null)) {
                onBroken.accept(entityIn);
                Item item = this.getItem();
@@ -288,7 +311,7 @@
    }
 
    public boolean isEffectiveOn(BlockState blockIn) {
-      return this.getItem().isEffectiveOn(blockIn);
+      return this.getItem().canHarvestBlock(this, blockIn);
    }
 
    public boolean useOnEntity(PlayerEntity playerIn, LivingEntity entityIn, Hand hand) {
@@ -296,7 +319,7 @@
    }
 
    public ItemStack copy() {
-      ItemStack itemstack = new ItemStack(this.getItem(), this.count);
+      ItemStack itemstack = new ItemStack(this.getItem(), this.count, this.serializeCaps());
       itemstack.setCooldown(this.getCooldown());
       if (this.tag != null) {
          itemstack.tag = this.tag.copy();
@@ -312,7 +335,7 @@
          if (stackA.tag == null && stackB.tag != null) {
             return false;
          } else {
-            return stackA.tag == null || stackA.tag.equals(stackB.tag);
+            return stackA.tag == null || stackA.tag.equals(stackB.tag) && stackA.areCapsCompatible(stackB);
          }
       } else {
          return false;
@@ -335,7 +358,7 @@
       } else if (this.tag == null && other.tag != null) {
          return false;
       } else {
-         return this.tag == null || this.tag.equals(other.tag);
+         return this.tag == null || this.tag.equals(other.tag) && this.areCapsCompatible(other);
       }
    }
 
@@ -641,6 +664,7 @@
          }
       }
 
+      net.minecraftforge.event.ForgeEventFactory.onItemTooltip(this, playerIn, list, advanced);
       return list;
    }
 
@@ -762,7 +786,7 @@
             }
          }
       } else {
-         multimap = this.getItem().getModifiers(equipmentSlot);
+         multimap = this.getItem().getAttributeModifiers(equipmentSlot, this);
       }
 
       return multimap;
@@ -901,4 +925,33 @@
    public boolean isFood() {
       return this.getItem().isFood();
    }
+
+   // FORGE START
+   public void deserializeNBT(CompoundTag nbt) {
+      final ItemStack itemStack = ItemStack.fromTag(nbt);
+      getStack().setTag(itemStack.getTag());
+      if (itemStack.capNBT != null) deserializeCaps(itemStack.capNBT);
+   }
+
+   /**
+    * Set up forge's ItemStack additions.
+    */
+   private void forgeInit() {
+      Item item = getItemRaw();
+      if (item != null) {
+         this.delegate = item.delegate;
+         net.minecraftforge.common.capabilities.ICapabilityProvider provider = item.initCapabilities(this, this.capNBT);
+         this.gatherCapabilities(provider);
+         if (this.capNBT != null) deserializeCaps(this.capNBT);
+      }
+   }
+
+   /**
+    * Internal call to get the actual item, not the delegate.
+    * In all other methods, FML replaces calls to this.item with the item delegate.
+    */
+   @Nullable
+   private Item getItemRaw() {
+       return this.item;
+   }
 }
