--- a/net/minecraft/item/Items.java
+++ b/net/minecraft/item/Items.java
@@ -11,10 +11,11 @@
 import net.minecraft.util.Identifier;
 import net.minecraft.sound.SoundEvents;
 import net.minecraft.util.registry.Registry;
+
 import net.minecraft.util.DyeColor;
 import net.minecraft.util.Rarity;
 
-
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Items {
    public static final Item AIR = register(Blocks.AIR, new AirBlockItem(Blocks.AIR, new Item.Settings()));
    public static final Item STONE = register(Blocks.STONE, ItemGroup.BUILDING_BLOCKS);
