--- a/net/minecraft/item/MilkBucketItem.java
+++ b/net/minecraft/item/MilkBucketItem.java
@@ -9,8 +9,8 @@
 import net.minecraft.util.ActionResult;
 import net.minecraft.util.Hand;
 import net.minecraft.world.World;
-import net.minecraft.util.UseAction;
 
+import net.minecraft.util.UseAction;
 
 public class MilkBucketItem extends Item {
    public MilkBucketItem(Item.Settings builder) {
@@ -18,6 +18,8 @@
    }
 
    public ItemStack finishUsing(ItemStack stack, World worldIn, LivingEntity entityLiving) {
+      if (!worldIn.isClient) entityLiving.curePotionEffects(stack); // FORGE - move up so stack.shrink does not turn stack into air
+
       if (entityLiving instanceof ServerPlayerEntity) {
          ServerPlayerEntity serverplayerentity = (ServerPlayerEntity)entityLiving;
          Criterions.CONSUME_ITEM.handle(serverplayerentity, stack);
@@ -47,4 +49,9 @@
       playerIn.setCurrentHand(handIn);
       return new TypedActionResult<>(ActionResult.SUCCESS, playerIn.getStackInHand(handIn));
    }
+
+   @Override
+   public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @javax.annotation.Nullable net.minecraft.nbt.CompoundTag nbt) {
+      return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+   }
 }
