--- a/net/minecraft/item/ShovelItem.java
+++ b/net/minecraft/item/ShovelItem.java
@@ -21,7 +21,7 @@
    protected static final Map<Block, BlockState> PATH_BLOCKSTATES = Maps.newHashMap(ImmutableMap.of(Blocks.GRASS_BLOCK, Blocks.GRASS_PATH.getDefaultState()));
 
    public ShovelItem(ToolMaterial tier, float attackDamageIn, float attackSpeedIn, Item.Settings builder) {
-      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder);
+      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder.addToolType(net.minecraftforge.common.ToolType.SHOVEL, tier.getMiningLevel()));
    }
 
    public boolean isEffectiveOn(BlockState blockIn) {
@@ -32,7 +32,7 @@
    public ActionResult useOnBlock(ItemUsageContext context) {
       World world = context.getWorld();
       BlockPos blockpos = context.getBlockPos();
-      if (context.getSide() != Direction.DOWN && world.getBlockState(blockpos.up()).isAir()) {
+      if (context.getSide() != Direction.DOWN && world.getBlockState(blockpos.up()).isAir(world, blockpos.up())) {
          BlockState blockstate = PATH_BLOCKSTATES.get(world.getBlockState(blockpos).getBlock());
          if (blockstate != null) {
             PlayerEntity playerentity = context.getPlayer();
