--- a/net/minecraft/client/render/entity/feature/ArmorBipedFeatureRenderer.java
+++ b/net/minecraft/client/render/entity/feature/ArmorBipedFeatureRenderer.java
@@ -39,4 +39,9 @@
    protected void method_4190(A model) {
       model.setVisible(false);
    }
+   
+   @Override
+   protected A getArmorModelHook(net.minecraft.entity.LivingEntity entity, net.minecraft.item.ItemStack itemStack, EquipmentSlot slot, A model) {
+      return net.minecraftforge.client.ForgeHooksClient.getArmorModel(entity, itemStack, slot, model);
+   }
 }
