--- a/net/minecraft/block/piston/PistonHandler.java
+++ b/net/minecraft/block/piston/PistonHandler.java
@@ -2,9 +2,7 @@
 
 import com.google.common.collect.Lists;
 import java.util.List;
-import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
-import net.minecraft.block.Blocks;
 import net.minecraft.block.PistonBlock;
 import net.minecraft.util.math.Direction;
 import net.minecraft.util.math.BlockPos;
@@ -51,7 +49,7 @@
       } else {
          for(int i = 0; i < this.movedBlocks.size(); ++i) {
             BlockPos blockpos = this.movedBlocks.get(i);
-            if (this.world.getBlockState(blockpos).getBlock() == Blocks.SLIME_BLOCK && !this.method_11538(blockpos)) {
+            if (this.world.getBlockState(blockpos).isStickyBlock() && !this.method_11538(blockpos)) {
                return false;
             }
          }
@@ -62,8 +60,7 @@
 
    private boolean tryMove(BlockPos origin, Direction facingIn) {
       BlockState blockstate = this.world.getBlockState(origin);
-      Block block = blockstate.getBlock();
-      if (blockstate.isAir()) {
+      if (world.isAir(origin)) {
          return true;
       } else if (!PistonBlock.isMovable(blockstate, this.world, origin, this.direction, false, facingIn)) {
          return true;
@@ -76,11 +73,10 @@
          if (i + this.movedBlocks.size() > 12) {
             return false;
          } else {
-            while(block == Blocks.SLIME_BLOCK) {
+            while(blockstate.isStickyBlock()) {
                BlockPos blockpos = origin.offset(this.direction.getOpposite(), i);
                blockstate = this.world.getBlockState(blockpos);
-               block = blockstate.getBlock();
-               if (blockstate.isAir() || !PistonBlock.isMovable(blockstate, this.world, blockpos, this.direction, false, this.direction.getOpposite()) || blockpos.equals(this.posFrom)) {
+               if (blockstate.isAir(this.world, blockpos) || !PistonBlock.isMovable(blockstate, this.world, blockpos, this.direction, false, this.direction.getOpposite()) || blockpos.equals(this.posFrom)) {
                   break;
                }
 
@@ -107,7 +103,7 @@
 
                   for(int l = 0; l <= k + i1; ++l) {
                      BlockPos blockpos2 = this.movedBlocks.get(l);
-                     if (this.world.getBlockState(blockpos2).getBlock() == Blocks.SLIME_BLOCK && !this.method_11538(blockpos2)) {
+                     if (this.world.getBlockState(blockpos2).isStickyBlock() && !this.method_11538(blockpos2)) {
                         return false;
                      }
                   }
@@ -116,7 +112,7 @@
                }
 
                blockstate = this.world.getBlockState(blockpos1);
-               if (blockstate.isAir()) {
+               if (blockstate.isAir(world, blockpos1)) {
                   return true;
                }
 
