--- a/net/minecraft/client/render/model/BakedQuad.java
+++ b/net/minecraft/client/render/model/BakedQuad.java
@@ -5,14 +5,26 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
+import net.minecraft;
+
 @OnlyIn(Dist.CLIENT)
-public class BakedQuad {
+public class BakedQuad implements net.minecraftforge.client.model.pipeline.IVertexProducer {
    protected final int[] vertexData;
    protected final int colorIndex;
    protected final Direction face;
    protected final Sprite sprite;
 
+   /**
+    * @deprecated Use constructor with the format argument.
+    */
+   @Deprecated
    public BakedQuad(int[] vertexDataIn, int tintIndexIn, Direction faceIn, Sprite spriteIn) {
+      this(vertexDataIn, tintIndexIn, faceIn, spriteIn, true, net.minecraft.client.render.VertexFormats.POSITION_COLOR_UV_LMAP);
+   }
+   
+   public BakedQuad(int[] vertexDataIn, int tintIndexIn, Direction faceIn, Sprite spriteIn, boolean applyDiffuseLighting, net.minecraft.client.render.VertexFormat format) {
+      this.format = format;
+      this.applyDiffuseLighting = applyDiffuseLighting;
       this.vertexData = vertexDataIn;
       this.colorIndex = tintIndexIn;
       this.face = faceIn;
@@ -38,4 +50,20 @@
    public Direction getFace() {
       return this.face;
    }
+   
+   protected final net.minecraft.client.render.VertexFormat format;
+   protected final boolean applyDiffuseLighting;
+   
+   @Override
+   public void pipe(net.minecraftforge.client.model.pipeline.IVertexConsumer consumer) {
+       net.minecraftforge.client.model.pipeline.LightUtil.putBakedQuad(consumer, this);
+   }
+   
+   public net.minecraft.client.render.VertexFormat getFormat() {
+       return format;
+   }
+   
+   public boolean shouldApplyDiffuseLighting() {
+       return applyDiffuseLighting;
+   }
 }
