--- a/net/minecraft/client/render/model/WeightedBakedModel.java
+++ b/net/minecraft/client/render/model/WeightedBakedModel.java
@@ -10,10 +10,10 @@
 import net.minecraft.util.WeightedPicker;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+
 import net.minecraft.client.render.model.json.ModelItemPropertyOverrideList;
 import net.minecraft.client.render.model.json.ModelTransformation;
 
-
 @OnlyIn(Dist.CLIENT)
 public class WeightedBakedModel implements BakedModel {
    private final int totalWeight;
@@ -33,6 +33,9 @@
    public boolean useAmbientOcclusion() {
       return this.defaultModel.useAmbientOcclusion();
    }
+   
+   @Override
+   public boolean isAmbientOcclusion(BlockState state) { return this.defaultModel.isAmbientOcclusion(state); }
 
    public boolean hasDepthInGui() {
       return this.defaultModel.hasDepthInGui();
