--- a/net/minecraft/entity/ai/goal/BreakDoorGoal.java
+++ b/net/minecraft/entity/ai/goal/BreakDoorGoal.java
@@ -4,7 +4,6 @@
 import net.minecraft.block.Block;
 import net.minecraft.entity.mob.MobEntity;
 import net.minecraft.world.Difficulty;
-import net.minecraft.world.GameRules;
 
 public class BreakDoorGoal extends DoorInteractGoal {
    private final Predicate<Difficulty> difficultySufficientPredicate;
@@ -29,7 +28,7 @@
    public boolean canStart() {
       if (!super.canStart()) {
          return false;
-      } else if (!this.mob.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+      } else if (!net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.mob.world, this.doorPos, this.mob)) {
          return false;
       } else {
          return this.isDifficultySufficient(this.mob.world.getDifficulty()) && !this.method_6256();
