--- a/net/minecraft/entity/ai/goal/MeleeAttackGoal.java
+++ b/net/minecraft/entity/ai/goal/MeleeAttackGoal.java
@@ -21,6 +21,8 @@
    private double targetZ;
    protected final int field_6504 = 20;
    private long field_19200;
+   private int failedPathFindingPenalty = 0;
+   private boolean canPenalize = false;
 
    public MeleeAttackGoal(MobEntityWithAi creature, double speedIn, boolean useLongMemory) {
       this.mob = creature;
@@ -41,6 +43,15 @@
          } else if (!livingentity.isAlive()) {
             return false;
          } else {
+           if (canPenalize) {
+               if (--this.field_6501 <= 0) {
+                  this.field_6509 = this.mob.getNavigation().findPathTo(livingentity, 0);
+                  this.field_6501 = 4 + this.mob.getRand().nextInt(7);
+                  return this.field_6509 != null;
+               } else {
+                  return true;
+               }
+            }
             this.field_6509 = this.mob.getNavigation().findPathTo(livingentity, 0);
             if (this.field_6509 != null) {
                return true;
@@ -92,6 +103,18 @@
          this.targetY = livingentity.getBoundingBox().minY;
          this.targetZ = livingentity.z;
          this.field_6501 = 4 + this.mob.getRand().nextInt(7);
+         if (this.canPenalize) {
+            this.field_6501 += failedPathFindingPenalty;
+            if (this.mob.getNavigation().getCurrentPath() != null) {
+               entity.ai.pathing.PathNode finalPathPoint = this.mob.getNavigation().getCurrentPath().getEnd();
+               if (finalPathPoint != null && livingentity.squaredDistanceTo(finalPathPoint.x, finalPathPoint.y, finalPathPoint.z) < 1)
+                  failedPathFindingPenalty = 0;
+               else
+                  failedPathFindingPenalty += 10;
+            } else {
+               failedPathFindingPenalty += 10;
+            }
+         }
          if (d0 > 1024.0D) {
             this.field_6501 += 10;
          } else if (d0 > 256.0D) {
