--- a/net/minecraft/entity/ai/goal/StepAndDestroyBlockGoal.java
+++ b/net/minecraft/entity/ai/goal/StepAndDestroyBlockGoal.java
@@ -11,7 +11,6 @@
 import net.minecraft.particle.ParticleTypes;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Vec3d;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.BlockView;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.ViewableWorld;
@@ -32,7 +31,7 @@
    }
 
    public boolean canStart() {
-      if (!this.stepAndDestroyMob.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+      if (!net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.stepAndDestroyMob.world, this.targetPos, this.stepAndDestroyMob)) {
          return false;
       } else if (this.cooldown > 0) {
          --this.cooldown;
