--- a/net\minecraft\enchantment\Enchantments.java
+++ b/net\minecraft\enchantment\Enchantments.java
@@ -3,6 +3,7 @@
 import net.minecraft.entity.EquipmentSlot;
 import net.minecraft.util.registry.Registry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Enchantments {
    private static final EquipmentSlot[] ALL_ARMOR = new EquipmentSlot[]{EquipmentSlot.field_6169, EquipmentSlot.field_6174, EquipmentSlot.field_6172, EquipmentSlot.field_6166};
    public static final Enchantment field_9111 = register("protection", new ProtectionEnchantment(Enchantment.Weight.field_9087, ProtectionEnchantment.Type.field_9138, ALL_ARMOR));
