--- a/net/minecraft/block/CactusBlock.java
+++ b/net/minecraft/block/CactusBlock.java
@@ -16,7 +16,7 @@
 import net.minecraft.world.ViewableWorld;
 import net.minecraft.world.World;
 
-public class CactusBlock extends Block {
+public class CactusBlock extends Block implements net.minecraftforge.common.IPlantable {
    public static final IntProperty AGE = Properties.AGE_15;
    protected static final VoxelShape COLLISION_SHAPE = Block.createCuboidShape(1.0D, 0.0D, 1.0D, 15.0D, 15.0D, 15.0D);
    protected static final VoxelShape OUTLINE_SHAPE = Block.createCuboidShape(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D);
@@ -27,6 +27,7 @@
    }
 
    public void onScheduledTick(BlockState state, World worldIn, BlockPos pos, Random random) {
+      if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent growing cactus from loading unloaded chunks with block update
       if (!state.canPlaceAt(worldIn, pos)) {
          worldIn.breakBlock(pos, true);
       } else {
@@ -39,6 +40,7 @@
 
             if (i < 3) {
                int j = state.get(AGE);
+               if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, blockpos, state, true)) {
                if (j == 15) {
                   worldIn.setBlockState(blockpos, this.getDefaultState());
                   BlockState blockstate = state.with(AGE, Integer.valueOf(0));
@@ -47,7 +49,8 @@
                } else {
                   worldIn.setBlockState(pos, state.with(AGE, Integer.valueOf(j + 1)), 4);
                }
-
+               net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
+               }
             }
          }
       }
@@ -82,8 +85,8 @@
          }
       }
 
-      Block block = worldIn.getBlockState(pos.down()).getBlock();
-      return (block == Blocks.CACTUS || block == Blocks.SAND || block == Blocks.RED_SAND) && !worldIn.getBlockState(pos.up()).getMaterial().isLiquid();
+      BlockState soil = worldIn.getBlockState(pos.down());
+      return soil.canSustainPlant(worldIn, pos.down(), Direction.UP, this) && !worldIn.getBlockState(pos.up()).getMaterial().isLiquid();
    }
 
    public void onEntityCollision(BlockState state, World worldIn, BlockPos pos, Entity entityIn) {
@@ -101,4 +104,14 @@
    public boolean canPlaceAtSide(BlockState state, BlockView worldIn, BlockPos pos, BlockPlacementEnvironment type) {
       return false;
    }
+
+   @Override
+   public net.minecraftforge.common.PlantType getPlantType(BlockView world, BlockPos pos) {
+      return net.minecraftforge.common.PlantType.Desert;
+   }
+
+   @Override
+   public BlockState getPlant(BlockView world, BlockPos pos) {
+      return getDefaultState();
+   }
 }
