--- a/net\minecraft\block\RedstoneOreBlock.java
+++ b/net\minecraft\block\RedstoneOreBlock.java
@@ -68,11 +68,11 @@
 
    public void onStacksDropped(BlockState state, World worldIn, BlockPos pos, ItemStack stack) {
       super.onStacksDropped(state, worldIn, pos, stack);
-      if (EnchantmentHelper.getLevel(Enchantments.field_9099, stack) == 0) {
-         int i = 1 + worldIn.random.nextInt(5);
-         this.dropExperience(worldIn, pos, i);
-      }
+   }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.WorldView world, BlockPos pos, int fortune, int silktouch) {
+      return silktouch == 0 ? 1 + RANDOM.nextInt(5) : 0;
    }
 
    @OnlyIn(Dist.CLIENT)
