--- a/net/minecraft/block/SaplingBlock.java
+++ b/net/minecraft/block/SaplingBlock.java
@@ -29,6 +29,7 @@
 
    public void onScheduledTick(BlockState state, World worldIn, BlockPos pos, Random random) {
       super.onScheduledTick(state, worldIn, pos, random);
+      if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
       if (worldIn.getLightLevel(pos.up()) >= 9 && random.nextInt(7) == 0) {
          this.generate(worldIn, pos, state, random);
       }
@@ -39,6 +40,7 @@
       if (state.get(STAGE) == 0) {
          worldIn.setBlockState(pos, state.cycle(STAGE), 4);
       } else {
+         if (!net.minecraftforge.event.ForgeEventFactory.saplingGrowTree(worldIn, rand, pos)) return;
          this.generator.generate(worldIn, pos, state, rand);
       }
 
