--- a/net/minecraft/block/SweetBerryBushBlock.java
+++ b/net/minecraft/block/SweetBerryBushBlock.java
@@ -47,8 +47,9 @@
    public void onScheduledTick(BlockState state, World worldIn, BlockPos pos, Random random) {
       super.onScheduledTick(state, worldIn, pos, random);
       int i = state.get(AGE);
-      if (i < 3 && random.nextInt(5) == 0 && worldIn.getLightLevel(pos.up(), 0) >= 9) {
+      if (i < 3 && worldIn.getLightLevel(pos.up(), 0) >= 9 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, random.nextInt(5) == 0)) {
          worldIn.setBlockState(pos, state.with(AGE, Integer.valueOf(i + 1)), 2);
+         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
       }
 
    }
