--- a/net/minecraft/block/TurtleEggBlock.java
+++ b/net/minecraft/block/TurtleEggBlock.java
@@ -19,7 +19,6 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.entity.EntityContext;
 import net.minecraft.util.shape.VoxelShape;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.BlockView;
 import net.minecraft.world.World;
 
@@ -145,7 +144,7 @@
       if (trampler instanceof TurtleEntity) {
          return false;
       } else {
-         return trampler instanceof LivingEntity && !(trampler instanceof PlayerEntity) ? worldIn.getGameRules().getBoolean(GameRules.MOB_GRIEFING) : true;
+         return trampler instanceof LivingEntity && !(trampler instanceof PlayerEntity) ? net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(worldIn, trampler) : true;
       }
    }
 }
