--- a/net/minecraft/block/VineBlock.java
+++ b/net/minecraft/block/VineBlock.java
@@ -19,7 +19,7 @@
 import net.minecraft.world.ViewableWorld;
 import net.minecraft.world.World;
 
-public class VineBlock extends Block {
+public class VineBlock extends Block implements net.minecraftforge.common.IShearable {
    public static final BooleanProperty UP = ConnectedPlantBlock.UP;
    public static final BooleanProperty NORTH = ConnectedPlantBlock.NORTH;
    public static final BooleanProperty EAST = ConnectedPlantBlock.EAST;
@@ -153,7 +153,7 @@
                worldIn.clearBlockState(pos, false);
             }
 
-         } else if (worldIn.random.nextInt(4) == 0) {
+         } else if (worldIn.random.nextInt(4) == 0 && worldIn.isAreaLoaded(pos, 4)) { // Forge: check area to prevent loading unloaded chunks
             Direction direction = Direction.random(random);
             BlockPos blockpos = pos.up();
             if (direction.getAxis().isHorizontal() && !state.get(getFacingProperty(direction))) {
@@ -328,4 +328,6 @@
    public static BooleanProperty getFacingProperty(Direction side) {
       return FACING_PROPERTIES.get(side);
    }
+
+   @Override public boolean isLadder(BlockState state, ViewableWorld world, BlockPos pos, net.minecraft.entity.LivingEntity entity) { return true; }
 }
