--- a/net/minecraft/world/gen/feature/BonusChestFeature.java
+++ b/net/minecraft/world/gen/feature/BonusChestFeature.java
@@ -20,7 +20,7 @@
    }
 
    public boolean generate(IWorld worldIn, ChunkGenerator<? extends ChunkGeneratorConfig> generator, Random rand, BlockPos pos, DefaultFeatureConfig config) {
-      for(BlockState blockstate = worldIn.getBlockState(pos); (blockstate.isAir() || blockstate.matches(BlockTags.LEAVES)) && pos.getY() > 1; blockstate = worldIn.getBlockState(pos)) {
+      for(BlockState blockstate = worldIn.getBlockState(pos); (blockstate.isAir(worldIn, pos) || blockstate.matches(BlockTags.LEAVES)) && pos.getY() > 1; blockstate = worldIn.getBlockState(pos)) {
          pos = pos.down();
       }
 
