--- a/net/minecraft/world/gen/feature/BushFeature.java
+++ b/net/minecraft/world/gen/feature/BushFeature.java
@@ -20,7 +20,7 @@
 
       for(int j = 0; j < 64; ++j) {
          BlockPos blockpos = pos.add(rand.nextInt(8) - rand.nextInt(8), rand.nextInt(4) - rand.nextInt(4), rand.nextInt(8) - rand.nextInt(8));
-         if (worldIn.isAir(blockpos) && (!worldIn.getDimension().isNether() || blockpos.getY() < 255) && blockstate.canPlaceAt(worldIn, blockpos)) {
+         if (worldIn.isAir(blockpos) && (!worldIn.getDimension().isNether() || blockpos.getY() < worldIn.getWorld().getDimension().getHeight()) && blockstate.canPlaceAt(worldIn, blockpos)) {
             worldIn.setBlockState(blockpos, blockstate, 2);
             ++i;
          }
