--- a/net/minecraft/world/gen/feature/JungleGrassFeature.java
+++ b/net/minecraft/world/gen/feature/JungleGrassFeature.java
@@ -23,7 +23,7 @@
    public boolean generate(IWorld worldIn, ChunkGenerator<? extends ChunkGeneratorConfig> generator, Random rand, BlockPos pos, DefaultFeatureConfig config) {
       BlockState blockstate = this.getGrass(rand);
 
-      for(BlockState blockstate1 = worldIn.getBlockState(pos); (blockstate1.isAir() || blockstate1.matches(BlockTags.LEAVES)) && pos.getY() > 0; blockstate1 = worldIn.getBlockState(pos)) {
+      for(BlockState blockstate1 = worldIn.getBlockState(pos); (blockstate1.isAir(worldIn, pos) || blockstate1.matches(BlockTags.LEAVES)) && pos.getY() > 0; blockstate1 = worldIn.getBlockState(pos)) {
          pos = pos.down();
       }
 
