--- a/net/minecraft/world/gen/feature/NetherSpringFeature.java
+++ b/net/minecraft/world/gen/feature/NetherSpringFeature.java
@@ -21,7 +21,7 @@
    public boolean generate(IWorld worldIn, ChunkGenerator<? extends ChunkGeneratorConfig> generator, Random rand, BlockPos pos, NetherSpringFeatureConfig config) {
       if (worldIn.getBlockState(pos.up()) != NETHERRACK) {
          return false;
-      } else if (!worldIn.getBlockState(pos).isAir() && worldIn.getBlockState(pos) != NETHERRACK) {
+      } else if (!worldIn.getBlockState(pos).isAir(worldIn, pos) && worldIn.getBlockState(pos) != NETHERRACK) {
          return false;
       } else {
          int i = 0;
