--- a/net/minecraft/world/gen/feature/SavannaTreeFeature.java
+++ b/net/minecraft/world/gen/feature/SavannaTreeFeature.java
@@ -23,7 +23,7 @@
    public boolean generate(Set<BlockPos> changedBlocks, ModifiableTestableWorld worldIn, Random rand, BlockPos position, MutableIntBoundingBox p_208519_5_) {
       int i = rand.nextInt(3) + rand.nextInt(3) + 5;
       boolean flag = true;
-      if (position.getY() >= 1 && position.getY() + i + 1 <= 256) {
+      if (position.getY() >= 1 && position.getY() + i + 1 <= worldIn.getMaxHeight()) {
          for(int j = position.getY(); j <= position.getY() + 1 + i; ++j) {
             int k = 1;
             if (j == position.getY()) {
@@ -38,7 +38,7 @@
 
             for(int l = position.getX() - k; l <= position.getX() + k && flag; ++l) {
                for(int i1 = position.getZ() - k; i1 <= position.getZ() + k && flag; ++i1) {
-                  if (j >= 0 && j < 256) {
+                  if (j >= 0 && j < worldIn.getMaxHeight()) {
                      if (!canTreeReplace(worldIn, blockpos$mutableblockpos.set(l, j, i1))) {
                         flag = false;
                      }
@@ -51,8 +51,8 @@
 
          if (!flag) {
             return false;
-         } else if (isNaturalDirtOrGrass(worldIn, position.down()) && position.getY() < 256 - i - 1) {
-            this.setToDirt(worldIn, position.down());
+         } else if (isSoil(worldIn, position.down(), getSapling()) && position.getY() < worldIn.getMaxHeight() - i - 1) {
+            this.setDirtAt(worldIn, position.down(), position);
             Direction direction = Direction.Type.HORIZONTAL.random(rand);
             int k2 = i - rand.nextInt(4) - 1;
             int l2 = 3 - rand.nextInt(3);
