--- a/net/minecraft/world/gen/feature/SpringFeature.java
+++ b/net/minecraft/world/gen/feature/SpringFeature.java
@@ -22,7 +22,7 @@
          return false;
       } else {
          BlockState blockstate = worldIn.getBlockState(pos);
-         if (!blockstate.isAir() && !Block.isNaturalStone(blockstate.getBlock())) {
+         if (!blockstate.isAir(worldIn, pos) && !Block.isNaturalStone(blockstate.getBlock())) {
             return false;
          } else {
             int i = 0;
