--- a/net/minecraft/entity/thrown/ThrownEnderpearlEntity.java
+++ b/net/minecraft/entity/thrown/ThrownEnderpearlEntity.java
@@ -83,6 +83,8 @@
       if (!this.world.isClient) {
          if (livingentity instanceof ServerPlayerEntity) {
             ServerPlayerEntity serverplayerentity = (ServerPlayerEntity)livingentity;
+               net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(serverplayerentity, this.x, this.y, this.z, 5.0F);
+               if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) { // Don't indent to lower patch size
             if (serverplayerentity.networkHandler.getConnection().isOpen() && serverplayerentity.world == this.world && !serverplayerentity.isSleeping()) {
                if (this.random.nextFloat() < 0.05F && this.world.getGameRules().getBoolean(GameRules.DO_MOB_SPAWNING)) {
                   EndermiteEntity endermiteentity = EntityType.ENDERMITE.create(this.world);
@@ -95,9 +97,10 @@
                   livingentity.stopRiding();
                }
 
-               livingentity.requestTeleport(this.x, this.y, this.z);
+               livingentity.requestTeleport(event.getTargetX(), event.getTargetY(), event.getTargetZ());
                livingentity.fallDistance = 0.0F;
-               livingentity.damage(DamageSource.FALL, 5.0F);
+               livingentity.damage(DamageSource.FALL, event.getAttackDamage());
+               }//Forge: End
             }
          } else if (livingentity != null) {
             livingentity.requestTeleport(this.x, this.y, this.z);
