--- a/net/minecraft/client/world/ClientWorld.java
+++ b/net/minecraft/client/world/ClientWorld.java
@@ -85,6 +85,8 @@
       this.setSpawnPos(new BlockPos(8, 64, 8));
       this.calculateAmbientDarkness();
       this.initWeatherGradients();
+      this.gatherCapabilities(dimension.initCapabilities());
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(this));
    }
 
    public void tick(BooleanSupplier hasTimeLeft) {
@@ -109,6 +111,7 @@
          Entity entity = this.globalEntities.get(i);
          this.tickEntity((p_217415_0_) -> {
             ++p_217415_0_.age;
+            if (p_217415_0_.canUpdate())
             p_217415_0_.tick();
          }, entity);
          if (entity.removed) {
@@ -156,6 +159,7 @@
             this.getProfiler().push(() -> {
                return Registry.ENTITY_TYPE.getId(p_217418_1_.getType()).toString();
             });
+            if (p_217418_1_.canUpdate())
             p_217418_1_.tick();
             this.getProfiler().pop();
          }
@@ -263,9 +267,11 @@
    }
 
    private void addEntityPrivate(int p_217424_1_, Entity p_217424_2_) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(p_217424_2_, this))) return;
       this.removeEntity(p_217424_1_);
       this.regularEntities.put(p_217424_1_, p_217424_2_);
       this.getChunkManager().getChunk(MathHelper.floor(p_217424_2_.x / 16.0D), MathHelper.floor(p_217424_2_.z / 16.0D), ChunkStatus.FULL, true).addEntity(p_217424_2_);
+      p_217424_2_.onAddedToWorld();
    }
 
    public void removeEntity(int eid) {
@@ -284,6 +290,7 @@
       }
 
       this.players.remove(p_217414_1_);
+      p_217414_1_.onRemovedFromWorld();
    }
 
    public void addEntitiesToChunk(WorldChunk p_217417_1_) {
@@ -409,6 +416,11 @@
    }
 
    public void playSound(@Nullable PlayerEntity player, double x, double y, double z, SoundEvent soundIn, SoundCategory category, float volume, float pitch) {
+      net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(player, soundIn, category, volume, pitch);
+      if (event.isCanceled() || event.getSound() == null) return;
+      soundIn = event.getSound();
+      category = event.getCategory();
+      volume = event.getVolume();
       if (player == this.client.player) {
          this.playSound(x, y, z, soundIn, category, volume, pitch, false);
       }
@@ -416,6 +428,11 @@
    }
 
    public void playSoundFromEntity(@Nullable PlayerEntity p_217384_1_, Entity p_217384_2_, SoundEvent p_217384_3_, SoundCategory p_217384_4_, float p_217384_5_, float p_217384_6_) {
+      net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(p_217384_1_, p_217384_3_, p_217384_4_, p_217384_5_, p_217384_6_);
+      if (event.isCanceled() || event.getSound() == null) return;
+      p_217384_3_ = event.getSound();
+      p_217384_4_ = event.getCategory();
+      p_217384_5_ = event.getVolume();
       if (p_217384_1_ == this.client.player) {
          this.client.getSoundManager().play(new EntityTrackingSoundInstance(p_217384_3_, p_217384_4_, p_217384_2_));
       }
