--- a/net\minecraft\block\entity\BeaconBlockEntity.java
+++ b/net\minecraft\block\entity\BeaconBlockEntity.java
@@ -11,7 +11,6 @@
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
-import net.minecraft.block.Stainable;
 import net.minecraft.container.BeaconContainer;
 import net.minecraft.container.BlockContext;
 import net.minecraft.container.Container;
@@ -112,8 +111,8 @@
       for(int i1 = 0; i1 < 10 && blockpos.getY() <= l; ++i1) {
          BlockState blockstate = this.world.getBlockState(blockpos);
          Block block = blockstate.getBlock();
-         if (block instanceof Stainable) {
-            float[] afloat = ((Stainable)block).getColor().getColorComponents();
+         float[] afloat = blockstate.getBeaconColorMultiplier(this.world, blockpos, getPos());
+         if (afloat != null) {
             if (this.field_19178.size() <= 1) {
                beacontileentity$beamsegment = new BeaconBlockEntity.BeamSegment(afloat);
                this.field_19178.add(beacontileentity$beamsegment);
@@ -184,8 +183,7 @@
 
          for(int k = beaconXIn - i; k <= beaconXIn + i && flag; ++k) {
             for(int l = beaconZIn - i; l <= beaconZIn + i; ++l) {
-               Block block = this.world.getBlockState(new BlockPos(k, j, l)).getBlock();
-               if (block != Blocks.field_10234 && block != Blocks.field_10205 && block != Blocks.field_10201 && block != Blocks.field_10085) {
+               if (!this.world.getBlockState(new BlockPos(k, j, l)).isBeaconBase(this.world, new BlockPos(k, j, l), getPos())) {
                   flag = false;
                   break;
                }
