--- a/net\minecraft\block\entity\BlockEntityType.java
+++ b/net\minecraft\block\entity\BlockEntityType.java
@@ -18,7 +18,7 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class BlockEntityType<T extends BlockEntity> {
+public class BlockEntityType<T extends BlockEntity> extends net.minecraftforge.registries.ForgeRegistryEntry<BlockEntityType<?>> {
    private static final Logger LOGGER = LogManager.getLogger();
    public static final BlockEntityType<FurnaceBlockEntity> field_11903 = create("furnace", BlockEntityType.Builder.create(FurnaceBlockEntity::new, Blocks.field_10181));
    public static final BlockEntityType<ChestBlockEntity> field_11914 = create("chest", BlockEntityType.Builder.create(ChestBlockEntity::new, Blocks.field_10034));
