--- a/net/minecraft/client/render/model/json/ModelItemPropertyOverrideList.java
+++ b/net/minecraft/client/render/model/json/ModelItemPropertyOverrideList.java
@@ -13,12 +13,13 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+
+import net.minecraft;
 import net.minecraft.client.render.model.BakedModel;
 import net.minecraft.client.render.model.ModelLoader;
 import net.minecraft.client.render.model.ModelRotation;
 import net.minecraft.client.render.model.UnbakedModel;
 
-
 @OnlyIn(Dist.CLIENT)
 public class ModelItemPropertyOverrideList {
    public static final ModelItemPropertyOverrideList EMPTY = new ModelItemPropertyOverrideList();
@@ -29,10 +30,14 @@
       this.models = Collections.emptyList();
    }
 
+   @Deprecated //Forge: Use IUnbakedModel, add texture getter
    public ModelItemPropertyOverrideList(ModelLoader p_i50984_1_, JsonUnbakedModel p_i50984_2_, Function<Identifier, UnbakedModel> p_i50984_3_, List<ModelItemOverride> p_i50984_4_) {
+      this(p_i50984_1_, (UnbakedModel)p_i50984_2_, p_i50984_3_, p_i50984_1_.spriteAtlas::getSprite, p_i50984_4_, net.minecraft.client.render.VertexFormats.POSITION_COLOR_UV_NORMAL);
+   }
+   public ModelItemPropertyOverrideList(ModelLoader p_i50984_1_, UnbakedModel p_i50984_2_, Function<Identifier, UnbakedModel> p_i50984_3_, Function<Identifier, net.minecraft.client.texture.Sprite> textureGetter, List<ModelItemOverride> p_i50984_4_, net.minecraft.client.render.VertexFormat format) {
       this.models = p_i50984_4_.stream().map((p_217649_3_) -> {
          UnbakedModel iunbakedmodel = p_i50984_3_.apply(p_217649_3_.getModelId());
-         return Objects.equals(iunbakedmodel, p_i50984_2_) ? null : p_i50984_1_.bake(p_217649_3_.getModelId(), ModelRotation.X0_Y0);
+         return Objects.equals(iunbakedmodel, p_i50984_2_) ? null : p_i50984_1_.getBakedModel(p_217649_3_.getModelId(), ModelRotation.X0_Y0, textureGetter, format);
       }).collect(Collectors.toList());
       Collections.reverse(this.models);
 
@@ -60,4 +65,8 @@
 
       return model;
    }
+
+   public com.google.common.collect.ImmutableList<ModelItemOverride> getOverrides() {
+      return com.google.common.collect.ImmutableList.copyOf(overrides);
+   }
 }
