--- a/net/minecraft/client/render/model/json/ModelVariant.java
+++ b/net/minecraft/client/render/model/json/ModelVariant.java
@@ -11,11 +11,11 @@
 import net.minecraft.util.Identifier;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
-import net.minecraft.client.render.model.ModelRotation;
 
+import net.minecraft.client.render.model.ModelRotation;
 
 @OnlyIn(Dist.CLIENT)
-public class ModelVariant implements ModelBakeSettings {
+public class ModelVariant implements ModelBakeSettings, net.minecraftforge.client.model.ISmartVariant {
    private final Identifier location;
    private final ModelRotation rotation;
    private final boolean uvLock;
@@ -32,6 +32,7 @@
       return this.location;
    }
 
+   @Deprecated
    public ModelRotation getRotation() {
       return this.rotation;
    }
