--- a/net/minecraft/client/network/packet/CustomPayloadS2CPacket.java
+++ b/net/minecraft/client/network/packet/CustomPayloadS2CPacket.java
@@ -8,7 +8,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class CustomPayloadS2CPacket implements Packet<ClientPlayPacketListener> {
+public class CustomPayloadS2CPacket implements Packet<ClientPlayPacketListener>, net.minecraftforge.fml.network.ICustomPacket<CustomPayloadS2CPacket> {
    public static final Identifier BRAND = new Identifier("brand");
    public static final Identifier DEBUG_PATH = new Identifier("debug/path");
    public static final Identifier DEBUG_NEIGHBORS_UPDATE = new Identifier("debug/neighbors_update");
