--- a/net\minecraft\world\border\WorldBorder.java
+++ b/net\minecraft\world\border\WorldBorder.java
@@ -134,6 +134,10 @@
       this.listeners.add(listener);
    }
 
+   public void removeListener(WorldBorderListener listener) {
+      this.listeners.remove(listener);
+   }
+
    public void setMaxWorldBorderRadius(int size) {
       this.maxWorldBorderRadius = size;
       this.area.onMaxWorldBorderRadiusChanged();
