--- a/net\minecraft\client\render\entity\model\SkeletonEntityModel.java
+++ b/net\minecraft\client\render\entity\model\SkeletonEntityModel.java
@@ -43,7 +43,7 @@
       this.rightArmPose = BipedEntityModel.ArmPose.field_3409;
       this.leftArmPose = BipedEntityModel.ArmPose.field_3409;
       ItemStack itemstack = entityIn.getStackInHand(Hand.field_5808);
-      if (itemstack.getItem() == Items.field_8102 && entityIn.isAttacking()) {
+      if (itemstack.getItem() instanceof net.minecraft.item.BowItem && entityIn.isAttacking()) {
          if (entityIn.getMainArm() == Arm.field_6183) {
             this.rightArmPose = BipedEntityModel.ArmPose.field_3403;
          } else {
@@ -57,7 +57,7 @@
    public void setAngles(T entityIn, float limbSwing, float limbSwingAmount, float ageInTicks, float netHeadYaw, float headPitch) {
       super.setAngles(entityIn, limbSwing, limbSwingAmount, ageInTicks, netHeadYaw, headPitch);
       ItemStack itemstack = entityIn.getMainHandStack();
-      if (entityIn.isAttacking() && (itemstack.isEmpty() || itemstack.getItem() != Items.field_8102)) {
+      if (entityIn.isAttacking() && (itemstack.isEmpty() || !(itemstack.getItem() instanceof net.minecraft.item.BowItem))) {
          float f = MathHelper.sin(this.handSwingProgress * (float)Math.PI);
          float f1 = MathHelper.sin((1.0F - (1.0F - this.handSwingProgress) * (1.0F - this.handSwingProgress)) * (float)Math.PI);
          this.rightArm.roll = 0.0F;
