--- a/net/minecraft/entity/mob/MagmaCubeEntity.java
+++ b/net/minecraft/entity/mob/MagmaCubeEntity.java
@@ -79,6 +79,7 @@
       Vec3d vec3d = this.getVelocity();
       this.setVelocity(vec3d.x, (double)(0.42F + (float)this.getSize() * 0.1F), vec3d.z);
       this.velocityDirty = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
    protected void swimUpward(Tag<Fluid> fluidTag) {
