--- a/net/minecraft/entity/EntityType.java
+++ b/net/minecraft/entity/EntityType.java
@@ -129,7 +129,7 @@
 
 import net.minecraft.entity.mob.MobEntity;
 
-public class EntityType<T extends Entity> {
+public class EntityType<T extends Entity> extends net.minecraftforge.registries.ForgeRegistryEntry<EntityType<?>> {
    private static final Logger LOGGER = LogManager.getLogger();
    public static final EntityType<AreaEffectCloudEntity> AREA_EFFECT_CLOUD = register("area_effect_cloud", EntityType.Builder.<AreaEffectCloudEntity>create(AreaEffectCloudEntity::new, EntityCategory.MISC).makeFireImmune().setDimensions(6.0F, 0.5F));
    public static final EntityType<ArmorStandEntity> ARMOR_STAND = register("armor_stand", EntityType.Builder.<ArmorStandEntity>create(ArmorStandEntity::new, EntityCategory.MISC).setDimensions(0.5F, 1.975F));
@@ -247,6 +247,12 @@
    private Identifier lootTableId;
    private final EntityDimensions dimensions;
 
+   private final java.util.function.Predicate<EntityType<?>> velocityUpdateSupplier;
+   private final java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier;
+   private final java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier;
+   private final java.util.function.BiFunction<net.minecraftforge.fml.network.FMLPlayMessages.SpawnEntity, World, T> customClientFactory;
+   private final net.minecraftforge.common.util.ReverseTagWrapper<EntityType<?>> reverseTags = new net.minecraftforge.common.util.ReverseTagWrapper<>(this, net.minecraft.tag.EntityTypeTags::getGeneration, net.minecraft.tag.EntityTypeTags::getContainer);
+
    private static <T extends Entity> EntityType<T> register(String key, EntityType.Builder<T> builder) {
       return Registry.register(Registry.ENTITY_TYPE, key, builder.build(key));
    }
@@ -259,7 +265,7 @@
       return Registry.ENTITY_TYPE.getOrEmpty(Identifier.tryParse(key));
    }
 
-   public EntityType(EntityType.EntityFactory<T> p_i51559_1_, EntityCategory p_i51559_2_, boolean p_i51559_3_, boolean p_i51559_4_, boolean p_i51559_5_, boolean p_i51559_6_, EntityDimensions p_i51559_7_) {
+   public EntityType(EntityType.EntityFactory<T> p_i51559_1_, EntityCategory p_i51559_2_, boolean p_i51559_3_, boolean p_i51559_4_, boolean p_i51559_5_, boolean p_i51559_6_, EntityDimensions p_i51559_7_, final java.util.function.Predicate<EntityType<?>> velocityUpdateSupplier, final java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier, final java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier, final java.util.function.BiFunction<net.minecraftforge.fml.network.FMLPlayMessages.SpawnEntity, World, T> customClientFactory) {
       this.factory = p_i51559_1_;
       this.category = p_i51559_2_;
       this.field_19423 = p_i51559_6_;
@@ -267,6 +273,10 @@
       this.summonable = p_i51559_4_;
       this.fireImmune = p_i51559_5_;
       this.dimensions = p_i51559_7_;
+      this.velocityUpdateSupplier = velocityUpdateSupplier;
+      this.trackingRangeSupplier = trackingRangeSupplier;
+      this.updateIntervalSupplier = updateIntervalSupplier;
+      this.customClientFactory = customClientFactory;
    }
 
    @Nullable
@@ -277,6 +287,7 @@
    @Nullable
    public T spawn(World worldIn, @Nullable CompoundTag compound, @Nullable Text customName, @Nullable PlayerEntity playerIn, BlockPos pos, SpawnType reason, boolean p_220342_7_, boolean p_220342_8_) {
       T t = this.create(worldIn, compound, customName, playerIn, pos, reason, p_220342_7_, p_220342_8_);
+      if (t instanceof minecraft.entity.mob.MobEntity && net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn((minecraft.entity.mob.MobEntity) t, worldIn, pos.getX(), pos.getY(), pos.getZ(), null, reason)) return null;
       worldIn.spawnEntity(t);
       return t;
    }
@@ -459,6 +470,9 @@
    }
 
    public int getMaxTrackDistance() {
+      return trackingRangeSupplier.applyAsInt(this);
+   }
+   private int defaultTrackingRangeSupplier() {
       if (this == PLAYER) {
          return 32;
       } else if (this == END_CRYSTAL) {
@@ -471,6 +485,9 @@
    }
 
    public int getTrackTickInterval() {
+      return updateIntervalSupplier.applyAsInt(this);
+   }
+   private int defaultUpdateIntervalSupplier() {
       if (this != PLAYER && this != EVOKER_FANGS) {
          if (this == EYE_OF_ENDER) {
             return 4;
@@ -491,6 +508,9 @@
    }
 
    public boolean alwaysUpdateVelocity() {
+      return velocityUpdateSupplier.test(this);
+   }
+   private boolean defaultVelocitySupplier() {
       return this != PLAYER && this != LLAMA_SPIT && this != WITHER && this != BAT && this != ITEM_FRAME && this != LEASH_KNOT && this != PAINTING && this != END_CRYSTAL && this != EVOKER_FANGS;
    }
 
@@ -498,12 +518,30 @@
       return tagIn.contains(this);
    }
 
+   public T customClientSpawn(net.minecraftforge.fml.network.FMLPlayMessages.SpawnEntity packet, World world) {
+      if (customClientFactory == null) return this.create(world);
+      return customClientFactory.apply(packet, world);
+   }
+
+   /**
+    * Retrieves a list of tags names this is known to be associated with.
+    * This should be used in favor of TagCollection.getOwningTags, as this caches the result and automatically updates when the TagCollection changes.
+    */
+   public java.util.Set<Identifier> getTags() {
+      return reverseTags.getTagNames();
+   }
+
    public static class Builder<T extends Entity> {
       private final EntityType.EntityFactory<T> factory;
       private final EntityCategory category;
       private boolean saveable = true;
       private boolean summonable = true;
       private boolean fireImmune;
+      private java.util.function.Predicate<EntityType<?>> velocityUpdateSupplier = EntityType::defaultVelocitySupplier;
+      private java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier = EntityType::defaultTrackingRangeSupplier;
+      private java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier = EntityType::defaultUpdateIntervalSupplier;
+      private java.util.function.BiFunction<net.minecraftforge.fml.network.FMLPlayMessages.SpawnEntity, World, T> customClientFactory;
+
       private boolean field_19424;
       private EntityDimensions size = EntityDimensions.changing(0.6F, 1.8F);
 
@@ -548,11 +586,35 @@
          return this;
       }
 
+      public EntityType.Builder<T> setUpdateInterval(int interval) {
+         this.updateIntervalSupplier = t->interval;
+         return this;
+      }
+
+      public EntityType.Builder<T> setTrackingRange(int range) {
+         this.trackingRangeSupplier = t->range;
+         return this;
+      }
+
+      public EntityType.Builder<T> setShouldReceiveVelocityUpdates(boolean value) {
+         this.velocityUpdateSupplier = t->value;
+         return this;
+      }
+
+      /**
+       * By default, entities are spawned clientside via {@link EntityType#create(World)}.
+       * If you need finer control over the spawning process, use this to get read access to the spawn packet.
+       */
+      public EntityType.Builder<T> setCustomClientFactory(java.util.function.BiFunction<net.minecraftforge.fml.network.FMLPlayMessages.SpawnEntity, World, T> customClientFactory) {
+         this.customClientFactory = customClientFactory;
+         return this;
+      }
+
       public EntityType<T> build(String id) {
          if (this.saveable) {
             try {
                Schemas.getFixer().getSchema(DataFixUtils.makeKey(SharedConstants.getGameVersion().getWorldVersion())).getChoiceType(TypeReferences.ENTITY_TREE, id);
-            } catch (IllegalStateException illegalstateexception) {
+            } catch (IllegalArgumentException illegalstateexception) { // Forge: fix catching wrong exception
                if (SharedConstants.isDevelopment) {
                   throw illegalstateexception;
                }
@@ -561,7 +623,7 @@
             }
          }
 
-         return new EntityType<>(this.factory, this.category, this.saveable, this.summonable, this.fireImmune, this.field_19424, this.size);
+         return new EntityType<>(this.factory, this.category, this.saveable, this.summonable, this.fireImmune, this.field_19424, this.size, velocityUpdateSupplier, trackingRangeSupplier, updateIntervalSupplier, customClientFactory);
       }
    }
 
