--- a/net/minecraft/client/render/item/ItemModels.java
+++ b/net/minecraft/client/render/item/ItemModels.java
@@ -31,7 +31,8 @@
 
    public Sprite getSprite(ItemStack stack) {
       BakedModel ibakedmodel = this.getModel(stack);
-      return (ibakedmodel == this.modelManager.getMissingModel() || ibakedmodel.isBuiltin()) && stack.getItem() instanceof BlockItem ? this.modelManager.getBlockStateMaps().getSprite(((BlockItem)stack.getItem()).getBlock().getDefaultState()) : ibakedmodel.getSprite();
+      // FORGE: Make sure to call the item overrides
+      return (ibakedmodel == this.modelManager.getMissingModel() || ibakedmodel.isBuiltin()) && stack.getItem() instanceof BlockItem ? this.modelManager.getBlockStateMaps().getSprite(((BlockItem)stack.getItem()).getBlock().getDefaultState()) : ibakedmodel.getItemPropertyOverrides().apply(ibakedmodel, stack, null, null).getSprite();
    }
 
    public BakedModel getModel(ItemStack stack) {
