--- a/net/minecraft/client/gui/screen/DeathScreen.java
+++ b/net/minecraft/client/gui/screen/DeathScreen.java
@@ -44,6 +44,7 @@
       }));
       ButtonWidget button = this.addButton(new ButtonWidget(this.width / 2 - 100, this.height / 4 + 96, 200, 20, s1, (p_213020_1_) -> {
          if (this.isHardcore) {
+            method_20373(true);
             this.minecraft.openScreen(new TitleScreen());
          } else {
             ConfirmScreen confirmscreen = new ConfirmScreen(this::method_20373, new TranslatableText("deathScreen.quit.confirm"), new LiteralText(""), I18n.translate("deathScreen.titleScreen"), I18n.translate("deathScreen.respawn"));
