--- a/net/minecraft/entity/passive/FoxEntity.java
+++ b/net/minecraft/entity/passive/FoxEntity.java
@@ -570,9 +570,10 @@
             }
          }
 
-         BlockState blockstate = this.world.getBlockState(new BlockPos(this.x, this.y - 0.2D - (double)this.prevYaw, this.z));
+         BlockPos pos = new BlockPos(this.x, this.y - 0.2D - (double)this.prevYaw, this.z);
+         BlockState blockstate = this.world.getBlockState(pos);
          if (!blockstate.isAir() && !this.isSilent()) {
-            BlockSoundGroup soundtype = blockstate.getSoundGroup();
+            BlockSoundGroup soundtype = blockstate.getSoundType(world, pos, this);
             this.world.playSound((PlayerEntity)null, this.x, this.y, this.z, soundtype.getStepSound(), this.getSoundCategory(), soundtype.getVolume() * 0.5F, soundtype.getPitch() * 0.75F);
          }
 
@@ -764,7 +765,7 @@
       }
 
       protected void eatSweetBerry() {
-         if (FoxEntity.this.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+         if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(FoxEntity.this.world, FoxEntity.this)) {
             BlockState blockstate = FoxEntity.this.world.getBlockState(this.targetPos);
             if (blockstate.getBlock() == Blocks.SWEET_BERRY_BUSH) {
                int i = blockstate.get(SweetBerryBushBlock.AGE);
