--- a/net\minecraft\entity\passive\RabbitEntity.java
+++ b/net\minecraft\entity\passive\RabbitEntity.java
@@ -490,7 +490,7 @@
 
       public boolean canStart() {
          if (this.cooldown <= 0) {
-            if (!this.rabbit.world.getGameRules().getBoolean(GameRules.field_19388)) {
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.rabbit.world, this.rabbit)) {
                return false;
             }
 
