--- a/net/minecraft/entity/passive/RabbitEntity.java
+++ b/net/minecraft/entity/passive/RabbitEntity.java
@@ -45,7 +45,6 @@
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.text.TranslatableText;
 import net.minecraft.world.LocalDifficulty;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.ViewableWorld;
 import net.minecraft.world.World;
@@ -497,7 +496,7 @@
 
       public boolean canStart() {
          if (this.cooldown <= 0) {
-            if (!this.rabbit.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.rabbit.world, this.rabbit)) {
                return false;
             }
 
