--- a/net\minecraft\entity\ai\brain\sensor\SensorType.java
+++ b/net\minecraft\entity\ai\brain\sensor\SensorType.java
@@ -4,7 +4,7 @@
 import net.minecraft.util.Identifier;
 import net.minecraft.util.registry.Registry;
 
-public class SensorType<U extends Sensor<?>> {
+public class SensorType<U extends Sensor<?>> extends net.minecraftforge.registries.ForgeRegistryEntry<SensorType<?>> {
    public static final SensorType<DummySensor> field_18465 = register("dummy", DummySensor::new);
    public static final SensorType<NearestLivingEntitiesSensor> field_18466 = register("nearest_living_entities", NearestLivingEntitiesSensor::new);
    public static final SensorType<NearestPlayersSensor> field_18467 = register("nearest_players", NearestPlayersSensor::new);
