--- a/net/minecraft/data/DataGenerator.java
+++ b/net/minecraft/data/DataGenerator.java
@@ -19,7 +19,7 @@
 
    public DataGenerator(Path output, Collection<Path> input) {
       this.output = output;
-      this.inputs = input;
+      this.inputs = Lists.newArrayList(input);
    }
 
    public Collection<Path> getInputs() {
@@ -53,6 +53,10 @@
       this.providers.add(provider);
    }
 
+   public void addInput(Path value) {
+      this.inputs.add(value);
+   }
+
    static {
       Bootstrap.initialize();
    }
